stablehlo/reference/Ops.cpp

```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
   return result;
 }
 
+Tensor eval(DotGeneralOp op, const Tensor &lhs, const Tensor &rhs) {
+  Tensor result(op.getType());
+  ArrayRef<int64_t> lhsShape = lhs.getType().getShape();
+  ArrayRef<int64_t> lhsBatchingDims =
+      op.getDotDimensionNumbers().getLhsBatchingDimensions();
+  ArrayRef<int64_t> rhsBatchingDims =
+      op.getDotDimensionNumbers().getRhsBatchingDimensions();
+  ArrayRef<int64_t> lhsContractingDims =
+      op.getDotDimensionNumbers().getLhsContractingDimensions();
+  ArrayRef<int64_t> rhsContractingDims =
+      op.getDotDimensionNumbers().getRhsContractingDimensions();
+  std::vector<int64_t> lhsNonContractingDims;
+  for (int64_t i = 0; i < lhs.getType().getRank(); ++i)
+    if (std::find(lhsBatchingDims.begin(), lhsBatchingDims.end(), i) ==
+            lhsBatchingDims.end() &&
+        std::find(lhsContractingDims.begin(), lhsContractingDims.end(), i) ==
+            lhsContractingDims.end())
+      lhsNonContractingDims.push_back(i);
+  std::vector<int64_t> rhsNonContractingDims;
+  for (int64_t i = 0; i < rhs.getType().getRank(); ++i)
+    if (std::find(rhsBatchingDims.begin(), rhsBatchingDims.end(), i) ==
+            rhsBatchingDims.end() &&
+        std::find(rhsContractingDims.begin(), rhsContractingDims.end(), i) ==
+            rhsContractingDims.end())
+      rhsNonContractingDims.push_back(i);
+  std::vector<int64_t> contractingDimSizes;
+  for (uint64_t i = 0; i < lhsContractingDims.size(); ++i)
+    contractingDimSizes.push_back(lhsShape[lhsContractingDims[i]]);
+  auto totalContractingSize = 1;
+  for (uint64_t i = 0; i < contractingDimSizes.size(); ++i)
+    totalContractingSize *= contractingDimSizes[i];
+  for (auto resIt = result.index_begin(); resIt != result.index_end();
+       ++resIt) {
+    Type resElTy =
+        op->getResultTypes().front().dyn_cast<ShapedType>().getElementType();
+    Element sum;
```

Comment:
Can we extract the computation to get 0-initialized sum in a lambda `initializeAccumulator`? That will make the code within the current loop more clear..